### ISSUE TITLE
Should users have to visit our web server?

### ISSUE THREAD

**Issue by nopdotcom** [jlund/streisand-discussions#31]
With fancier client installation methods on the horizon, should we still expect users to visit our HTTPS server first?

Apple `mobileconfig` profiles can contain multiple payloads. The bare-mobileconfig branch is called “bare" because it only ships a VPN definition.  A mobileconfig could also ship the server SSL certs more conveniently than the current method. (The cert mobileconfig would be a data: URL in streisand.html.) While we’re at it, we can also deploy client SSL certs, if we wanted to authenticate clients that way.

The fastest path to an automatic Windows L2TP setup tool may be either an NSIS “installer” (which would be generated for each Streisand server) or PowerShell. These too have a chance to drop SSL certs on the machine.

So for some users, the web site may be just a speedbump to getting to the juicy mobileconfig/installer files inside. Instead of sharing streisand.html, they could share the config files instead. But is that a good behavior to encourage?

One path to server migration is to post instructions on old-streisand (#30) but if users can’t/won’t visit https://old-streisand, then instructions there won’t help.

**Comment by cpu** [comment]
Interesting question. I haven't given it that much thought & don't know what the best answer would be.

Prior to working on Streisand I was dabbling with my own hacked up fork in which I removed the gateway in favour of placing a directory structure of required secrets/configs on the provisioning machine along with accompanying `.zip` and `.tar.gz` archives. That might be something to consider in addition to the gateway. I think the Gateway adds value and while iOS/Windows clients may have fancier provisioning support with `mobileconfig` files (etc) I suspect there will always be outliers (e.g. Wireguard, the Tor relay).

Food for thought!

**Comment by ghost** [comment]
Restricting it to the `mobileconfig` for iOS/macOS devices would be an overbearing action. Including the stuff for other OSes as well will render the webserver itself superfluous.
The server can be updated so that the mirrored software is current, too. A static file, which was handed over can be obsolete quite fast.
I'd stick to the current scheme (handing over visiting instructions, software download from the server).

**Comment by nickolasclarke** [comment]
My thoughts on the issue from an old thread with @jlund: 

> The gateway may be used once, but since it requires a SSL cert to be added to access it, and still gets complaints about a possibly unsecure connection, it is a bit beyond the layman still. Personally, I often just scp the sites-available folder w/out the mirrored clients to my machine quickly so I dont need to worry about connecting to the gateway. All that said, I think the gateway is still valuable to have, but I think getting access to the juicy bits, i.e. the configuration details, could be made more accessible quickly. I've not dig into this, but are the gateway markdown files generated on the client side, or the remote server side? If I understand correctly, all the details used to generate that markdown are also accessible to the ansible client running the playbooks, so co-generating a copy on both sides (sans clients perhaps) could be a easy solution to this, especially if it can be done in parallel.

basically I'd be a fan of being able to email or getting a copy of the full gateway site locally on the machine I am generating the server on. The gateway at that point just becomes a backup if I don't have access in the future. Currently the gateway and the hoops required to get onto it are just a lot of friction.

**Comment by cpu** [comment]
> basically I'd be a fan of being able to email or getting a copy of the full gateway site locally on the machine I am generating the server on. 

+1 - I think having a .zip of the gateway site automatically downloaded to the provisioning machine is sensible.

**Comment by nopdotcom** [comment]
This is what I was planning on doing:

# *This describes an imaginary version of Streisand * #

This is just a proposal.

# Getting your access information #

tl;dr: grab the files out of `generated-docs`, and give `streisand.html` to people you want to use your server.  Check the certificate. There's a way to get `generated-docs` without ssh.

## Why you need it

Each Streisand server controls access to itself. For VPN protocols, your Streisand gateway (web server) has documentation for each protocol; the documents give the needed configuration files and passphrases.

To get this information, you need to visit your Streisand web server. Here's how.

### Server identity

When you connect to the web server, how do you know you are connected securely to *your* Streisand server? 

A third-party could set up an imposter server. It would read requests from you and decode them. The imposter server would know what you were doing, and be able to censor results.

Because of this kind of threat, each Streisand server uses an *HTTPS certificate* to identify itself. But how do you get that certificate to verify it? You can't ask the server for it, since a hostile server could just lie.

Streisand generates an HTML instruction file. It contains an installable copy of the certificate. If you can't install that certificate, the instructions detail how to verifying the server's certificate against the instruction file.

### Access control ###

Streisand servers are generally private. They don't have user accounts; instead, knowledge of a long passphrase or other secret is sufficient to use Streisand resources.

(Why no user accounts? For simplicity, but also from the principle "Don't log anything you don't need.")

How do you get access to these secrets? Each protocol's instructions on the web server has its secrets, but how do you connect to the web server in the first place?

First, make sure you're connecting to the server you think you're connecting to: check the certificates. See above.

The same HTML file that contains your Streisand web server certificate also contains the passphrase for the web server. In other words, you need to know a secret from this document in order to do anything with your Streisand server. So you need that document, and so does anyone else using this Streisand server.

## Downloading ##

### Using `scp`/`rsync` ###

We all know this. Requires knowledge of ssh.

### Use a temporary web server ###

When the Streisand builder generates the access documents, it creates an encrypted ZIP file containing them too. After you've set up a Streisand server, you can use the `tbd` command on the builder to show the ZIP file's password and hash. `tbd2` runs a temporary web server that only serves that one ZIP file. 

### Verifying hashes ###

When you've downloaded the file, you should verify the ZIP file's hash; that way, you know nobody has tampered with it in transit. We assume your connection to the builder is secure, so you can trust the hash you see from the `tbd` command.

#### Microsoft Windows ####

Most Windows installations contain Microsoft PowerShell, which has a file hash command. To use it:

1. From the Windows menu, either search for or *Run* "powershell".
1. Change to the directory with the downloaded file. This command will take you to the most likely place: `cd $env:userprofile/downloads`
1. Get the file hash. If your ZIP file is named `streisand.zip`, then run the command `get-filehash -a sha256 yourfilename.zip`, which returns the SHA256 checksum.

#### Linux ####

1. Open a terminal window.
1. Change to the directory with the downloaded file. This will probably take you to the right place: `cd ~/Downloads`
1. If your ZIP file is named `streisand.zip`, run `sha256sum streisand.zip`

#### OS X ####

1. Open Terminal.app, open a terminal window
1. Change to the directory with the downloaded file. This will probably take you to the right place: `cd ~/Downloads`
1. If your ZIP file is named `streisand.zip`, run `shasum -a 256 streisand.zip`
